eval: remove deprecatedContext field from Context

yuzefovich opened the pull request:
This PR completes the effort to fully remove `context.Context` stored in the `eval.Context`. The only remaining usage has been when evaluating the placeholder expressions, and this PR plumbs `context.Context` as an explicit argument there - by introducing it as an argument to most methods of `tree.Datum` interface. Most of the changes in this PR have been mechanical - only the fourth commit contains most of the non-trivial plumbing and some new logic.

Epic: None

yuzefovich committed:
tree: rename CompareContext parameter ctx to cmpCtx

This is done in preparation of introducing `context.Context` argument to
many methods.

Release note: None

yuzefovich committed:
tree: add context argument to CompareError method

This commit begins plumbing of `context.Context` argument to most
methods of the `tree.Datum` interface in an effort to remove the stored
context from `eval.Context` struct. The amount of new plumbing in this
commit is relatively small. For now all `Compare` implementations
hard-code `context.Background()` usage and that will be replaced in the
following commit.

There is one spot where we introduce `context.TODO` that will not be
removed immediately - `coldataext.CompareDatum`. This method is used
extensively in the vectorized engine, and it'll require lots of
plumbing. I actually have another WIP PR that does most of that plumbing
(for a different reason), so I plan to remove this usage of
`context.TODO` after the other change lands.

Release note: None

yuzefovich committed:
sql: switch all usages of Datum.Compare to CompareError

This commit replaces all usages of `tree.Datum.Compare` method with
`CompareError`. This achieves two purposes:
- making it clear that the method might result in an error so now it is
the caller's job to explicitly decide on how to handle errors. In many
cases we still panic as we did before (for example, in the optimizer
this is the expected way of handling errors).
- propagating the `context.Context` argument to all the callsites. This
required quite a bit of plumbing to get done. Most of the new plumbing
is in the optimizer packages.

Release note: None

yuzefovich committed:
tree,opt: remove a couple of unused methods

This commit removes IsDistinctFrom which is only used in a single unit
test and HasRemoteSpans that isn't used at all.

Release note: None

yuzefovich committed:
tree: remove Compare and rename CompareError to Compare

All usages of old `Compare` have been switched to `CompareError`
instead. This allows us to rename `CompareError` to `Compare` to clean
up the interface.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125715)
<!-- Reviewable:end -->

yuzefovich committed:
tree: add context argument to remaining methods of Datum interface

This commit plumbs `context.Context` argument to `Prev`, `IsMin`,
`Next`, `IsMax`, `Max`, and `Min` methods of the `Datum` interface.

Release note: None

yuzefovich committed:
eval: remove deprecatedContext field from Context

This commit completes the deprecation of the `context.Context` stored in
the `eval.Context`. Its last remaining usage was in
`MustGetPlaceholderValue`, and now all callers of that method have
explicit `context.Context` in scope, we can safely pass it as
a parameter. Apart from being a nice cleanup, this commit paves the way
towards reducing some allocations (namely, we will no longer require
creating copies of the eval context for processors since mutating the
deprecated context was the only modification for many of them - this
will be done separately).

Release note: None

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice job with all the plumbing :potable_water: 

> Reviewed 3 of 3 files at r1, 2 of 2 files at r2, 52 of 52 files at r3, 106 of 106 files at r4, 68 of 68 files at r5, 15 of 15 files at r6, 14 of 14 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125715#-:-O-cAaM84ozvtw81PPLf:bj564z2)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich)

___
*[`pkg/cli/statement_bundle.go` line 354 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125715#-O-c6QxT-H6iAIZMtV6q:-O-c6QxT-H6iAIZMtV6r:b-12dpg8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4816d8947befda6b214cfac838c6d2d18c7d779/pkg/cli/statement_bundle.go#L354)):*
> ```Go
> 				} else if cmp, err := maxUpperBound.CompareError(ctx, &evalCtx, datum); err != nil {
> 					panic(err)
> 				} else if cmp < 0 {
> ```

TIL that a var like `cmp` is in scope in a later `else if`. I had assumed it was only in scope after the `;` and in that branch's block. Cool!


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for doing this!

> Reviewed 3 of 3 files at r1, 2 of 2 files at r2, 52 of 52 files at r3, 106 of 106 files at r4, 68 of 68 files at r5, 15 of 15 files at r6, 14 of 14 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125715#-:-O-cGbQR0jqfUgzPvSwn:b-pt8m11)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Thanks for the reviews!

bors r+
